evm/src/cpu/kernel/asm/curve/bls381/util.asm

```diff
@@ -44,10 +44,80 @@ global test_add_fp381:
     %add_fp381
     %jump(0xdeadbeef)
 
+global test_sub_fp381:
+    %sub_fp381
+    %jump(0xdeadbeef)
+
 global test_mul_fp381:
     %mul_fp381
     %jump(0xdeadbeef)
 
-global test_sub_fp381:
+
+%macro add_fp381_2
+    // stack: x_re, x_im, y_re, y_im
+    %stack (x_re: 2, x_im: 2, y_re: 2, y_im: 2) -> (y_im, x_im, y_re, x_re)
+    // stack: y_im, x_im, y_re, x_re
+    %add_fp381
+    // stack:       z_im, y_re, x_re
+    %stack (z_im: 2, y_re: 2, x_re: 2) -> (x_re, y_re, z_im)
+    // stack:       x_re, y_re, z_im
+    %add_fp381
+    // stack:             z_re, z_im
+%endmacro
+
+%macro sub_fp381_2
+    // stack: x_re, x_im, y_re, y_im
+    %stack (x_re: 2, x_im: 2, y_re: 2, y_im: 2) -> (x_im, y_im, y_re, x_re)
+    // stack: x_im, y_im, y_re, x_re
+    %sub_fp381
+    // stack:       z_im, y_re, x_re
+    %stack (z_im: 2, y_re: 2, x_re: 2) -> (x_re, y_re, z_im)
+    // stack:       x_re, y_re, z_im
     %sub_fp381
+    // stack:             z_re, z_im
+%endmacro
+
+global test_add_fp381_2:
+    %add_fp381_2
+    %jump(0xdeadbeef)
+
+global test_sub_fp381_2:
+    %sub_fp381_2
     %jump(0xdeadbeef)
+
+
+global mul_fp381_2:
+    // stack:                          x_re, x_im, y_re, y_im, jumpdest
```

Comment:
The `stack:` comments are helpful, but you can make them clearer by somehow denoting that `x_re`, ..., `y_im` each take up two slots. Otherwise it looks like `DUP4` is copying `y_im`.